internal/backends/python/requirements.go

```diff
@@ -0,0 +1,167 @@
+//
+// Specification reference:
+//
+//   https://pip.pypa.io/en/stable/reference/requirements-file-format/#example
+
+var pep345Name = `(?:[A-Z0-9]|[A-Z0-9][A-Z0-9._-]*[A-Z0-9])`
+var pep440VersionSpec = `(?:(?:~=|!=|===|==|>=|<=|>|<)\s*[^, ]+)+`
+var extrasSpec = `\[(` + pep345Name + `(?:\s*,\s*` + pep345Name + `)*)\]`
+var matchPackageAndSpec = regexp.MustCompile(`(?i)^\s*(` + pep345Name + `)\s*` + `((?:` + extrasSpec + `)?\s*` + pep440VersionSpec + `)\s*$`)
+
+// Global options:
+//
+//	https://pip.pypa.io/en/stable/reference/requirements-file-format/#global-options
+//
+// NB: We explicitly ignore -r/--requirement and -c/--constraint here, since we handle them separately
+var knownFlags = map[string]bool{
+	`-i`:                true,
+	`--index-url`:       true,
+	`--extra-index-url`: true,
+	`--no-index`:        true,
+	`-e`:                true,
+	`--editable`:        true,
+	`-f`:                true,
+	`--find-links`:      true,
+	`--no-binary`:       true,
+	`--only-binary`:     true,
+	`--prefer-binary`:   true,
+	`--require-hashes`:  true,
+	`--pre`:             true,
+	`--trusted-host`:    true,
+	`--use-feature`:     true,
+}
+
+type PipFlag string
+type Constraints map[api.PkgName]api.PkgSpec
+
+func findPackage(line string) (*api.PkgName, *api.PkgSpec, bool) {
+	var name *api.PkgName
+	var spec *api.PkgSpec
+
+	var found bool
+
+	matches := matchPackageAndSpec.FindSubmatch([]byte(line))
+	if len(matches) > 1 {
+		_name := api.PkgName(string(matches[1]))
+		name = &_name
+		found = true
+	}
+	if len(matches) > 2 {
+		_spec := api.PkgSpec(string(matches[2]))
+		spec = &_spec
+	}
+	return name, spec, found
+}
+
+func recurseRequirementsTxt(depth int, path string, sofar map[api.PkgName]api.PkgSpec, constraints Constraints) ([]PipFlag, map[api.PkgName]api.PkgSpec, Constraints) {
+	// Perhaps this can be lifted, but sensibly attempt to protect ourselves
+	if depth > 10 {
+		util.Die("Too many -r redirects in %s", path)
+	}
+
+	var flags []PipFlag
+
+	handle, err := os.Open(path)
+	if err != nil {
+		util.Die("%s", err.Error())
+	}
+	defer handle.Close()
+
+	scanner := bufio.NewScanner(handle)
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+
+		// Separate out comments
+		segments := strings.SplitN(line, "#", 1)
+		if len(segments) == 2 {
+			line = strings.TrimSpace(segments[0])
+			// comment = strings.TrimSpace(segments[1])
+		}
+
+		if line == "" {
+			// Skip blank lines
+		} else if name, spec, found := findPackage(line); found {
+			// Found a package!
+			sofar[*name] = *spec
+		} else if nextfile, found := strings.CutPrefix(line, "-r "); found {
+			// # It is possible to refer to other requirement files...
+			// -r other-requirements.txt
+			var newFlags []PipFlag
+			newFlags, sofar, constraints = recurseRequirementsTxt(depth+1, strings.TrimSpace(nextfile), sofar, constraints)
+			flags = append(flags, newFlags...)
+		} else if _, found := strings.CutPrefix(line, "-c "); found {
```

Comment:
TIL about constraints.

Reply:
Oh man, reading the requirements.txt spec, yikes.

Reply:
python always makes me feel like i'm tripping